File: backends/python-fastapi/app/repositories/invitation_repo.py
import secrets
from dataclasses import dataclass
from typing import Optional
from sqlalchemy.orm import Session
from app.models.models import Invitation


@dataclass
class InvitationResult:
    token: str
    url: str
    display_url: str


def _build_result(token: str, frontend_url: str) -> InvitationResult:
    url = f"{frontend_url}/invitation/{token}"
    display_url = url.replace("https://", "").replace("http://", "")
    return InvitationResult(token=token, url=url, display_url=display_url)


class InvitationRepository:
    def __init__(self, db: Session, frontend_url: str):
        self.db = db
        self.frontend_url = frontend_url

    def get_current(self) -> Optional[InvitationResult]:
        inv = self.db.query(Invitation).order_by(Invitation.id.desc()).first()
        if inv is None:
            return None
        return _build_result(inv.token, self.frontend_url)

    def issue(self) -> InvitationResult:
        token = secrets.token_hex(16)
        inv = Invitation(token=token)
        self.db.add(inv)
        self.db.commit()
        return _build_result(token, self.frontend_url)

    def find_by_token(self, token: str) -> Optional[Invitation]:
        return self.db.query(Invitation).filter(Invitation.token == token).first()

File: backends/go-gin/internal/repository/invitation_repo.go
package repository

import (
	"authorization-go/internal/model"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"time"

	"gorm.io/gorm"
)

type InvitationRepository struct {
	db          *gorm.DB
	frontendURL string
}

func NewInvitationRepository(db *gorm.DB, frontendURL string) *InvitationRepository {
	return &InvitationRepository{db: db, frontendURL: frontendURL}
}

type InvitationResult struct {
	Token      string
	URL        string
	DisplayURL string
}

func (r *InvitationRepository) GetCurrent() (*InvitationResult, error) {
	var inv model.Invitation
	if err := r.db.Order("id DESC").First(&inv).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return r.buildResult(inv.Token), nil
}

func (r *InvitationRepository) Issue() (*InvitationResult, error) {
	token, err := generateToken()
	if err != nil {
		return nil, err
	}

	now := time.Now()
	inv := model.Invitation{
		Token:     token,
		CreatedAt: now,
		UpdatedAt: now,
	}
	if err := r.db.Create(&inv).Error; err != nil {
		return nil, err
	}
	return r.buildResult(token), nil
}

func (r *InvitationRepository) FindByToken(token string) (*InvitationResult, error) {
	var inv model.Invitation
	if err := r.db.Where("token = ?", token).First(&inv).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return r.buildResult(inv.Token), nil
}

func (r *InvitationRepository) buildResult(token string) *InvitationResult {
	url := fmt.Sprintf("%s/register?token=%s", r.frontendURL, token)
	displayURL := url
	if len(displayURL) > 50 {
		displayURL = displayURL[:20] + "..." + displayURL[len(displayURL)-20:]
	}
	return &InvitationResult{Token: token, URL: url, DisplayURL: displayURL}
}

func generateToken() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}
